platform/qemu.go

```diff
@@ -34,80 +45,208 @@ type Disk struct {
+	swtpm     exec.Cmd
+}
+
+func (inst *QemuInstance) Pid() int {
+	return inst.qemu.Pid()
+}
+
+func (inst *QemuInstance) Destroy() {
+	if inst.swtpmTmpd != "" {
+		if inst.swtpm != nil {
+			inst.swtpm.Kill() // Ignore errors
+		}
+		if err := os.RemoveAll(inst.swtpmTmpd); err != nil {
+			plog.Errorf("Error removing swtpm dir: %v", err)
+		}
 	}
-	if d.Size != "" && d.BackingFile != "" {
-		return nil, ErrBothSizeAndFile
+	if inst.qemu != nil {
+		if err := inst.qemu.Kill(); err != nil {
+			plog.Errorf("Error killing qemu instance %v: %v", inst.Pid(), err)
+		}
 	}
+}
+
+// QemuBuilder is a configurator that can then create a qemu instance
+type QemuBuilder struct {
+	Board string
+
+	// Config is a path to Ignition configuration
+	Config string
+
+	Memory     int
+	Processors int
+	Uuid       string
+	Swtpm      bool
+	Pdeathsig  bool
+	Argv       []string
 
-	if d.Size != "" {
-		return setupDisk(d.ConfPath, d.Size)
-	} else {
-		return setupDiskFromFile(d.BackingFile, d.ConfPath)
+	primaryDiskAdded bool
+
+	finalized bool
+	diskId    uint
+	fds       []*os.File
+}
+
+func NewBuilder(board, config string) *QemuBuilder {
+	ret := QemuBuilder{
+		Board:     board,
+		Config:    config,
+		Swtpm:     true,
+		Pdeathsig: true,
+		Argv:      []string{},
 	}
+	return &ret
 }
 
-// Create a nameless temporary qcow2 image file backed by a raw image.
-func setupDiskFromFile(imageFile string, confPath string) (*os.File, error) {
-	// a relative path would be interpreted relative to /tmp
-	backingFile, err := filepath.Abs(imageFile)
+// AddFd appends a file descriptor that will be passed to qemu,
+// returning a "/dev/fdset/<num>" argument that one can use with e.g.
+// -drive file=/dev/fdset/<num>.
+func (builder *QemuBuilder) AddFd(fd *os.File) string {
+	set := len(builder.fds) + 1
+	builder.fds = append(builder.fds, fd)
+	return fmt.Sprintf("/dev/fdset/%d", set)
+}
+
+// AddQcow2Disk adds a disk image from a file descriptor,
+// mounted read-write, formatted qcow2.
+func (builder *QemuBuilder) AddQcow2DiskFd(fd *os.File, options []string) {
+	opts := ""
+	if len(options) > 0 {
+		opts = "," + strings.Join(options, ",")
+	}
+	fdset := builder.AddFd(fd)
+	id := fmt.Sprintf("d%d", builder.diskId)
+	builder.diskId += 1
+	builder.Append("-drive", fmt.Sprintf("if=none,id=%s,format=qcow2,file=%s,auto-read-only=off", id, fdset),
+		"-device", virtio(builder.Board, "blk", fmt.Sprintf("drive=%s%s", id, opts)))
+}
+
+func (builder *QemuBuilder) ConsoleToFile(path string) {
+	builder.Append("-display", "none", "-chardev", "file,id=log,path="+path, "-serial", "chardev:log")
+}
+
+func (builder *QemuBuilder) EnableUsermodeNetworking(forwardedPort uint) {
+	var forward string
+	if forwardedPort != 0 {
+		forward = fmt.Sprintf(",hostfwd=tcp:127.0.0.1:0-:%d", forwardedPort)
+	}
+	builder.Append("-netdev", "user,id=eth0"+forward, "-device", virtio(builder.Board, "net", "netdev=eth0"))
+}
+
+func resolveBackingFile(backingFile string) (string, error) {
```

Comment:
minor nit: I'd prefer to have the external function & class method definitions not be interleaved

Reply:
Hmm...what do you mean by "external"?  These aren't exported.  Is this more just like "order functions before methods"?

If so I think that generally makes sense...will do.

Reply:
Hmm OK so I'll be honest, bouncing back and forth between shell/Golang/Rust/C/Python (not to mention Make, spec files etc.) so often makes it harder for me to internalize style.  I think my Golang, Rust and C have tended to intermix functions and methods.

What I think happens commonly is that a helper function gets extracted from a method.  And it's most natural to place that helper function just above the method.

Files often end up hence being ordered at least partially by their evolution over time.

From a quick survey of the mantle code and other Golang projects...it certainly doesn't seem like there's a consistent "functions before methods" - do you have a specific example or reference?  (Again, sorry if this is well known, I just don't *live* in Golang)

I think we could probably clean up the ordering here some though for sure...but note doing so would inflate the diff more.

Reply:
OK, I added another commit which tries to clean things up here some more.

Reply:
I think the guestfish injection stuff should probably be split to a separate file, but...trying to be incremental.

Reply:
> I think the guestfish injection stuff should probably be split to a separate file, but...trying to be incremental.

That's fine!